src/main/kotlin/com/rapatao/projects/ruleset/engine/Evaluator.kt

```diff
@@ -77,5 +80,21 @@ class Evaluator(
         return allMatch
     }
 
-    private fun Expression.processExpression(context: EvalContext): Boolean = context.process(this)
+    private fun Expression.processExpression(context: EvalContext): Boolean {
+        return usingFailureWrapper(this.onFailure) {
+            context.process(this)
+        }
+    }
+
+    private fun usingFailureWrapper(onFailure: OnFailure, block: () -> Boolean): Boolean {
```

Comment:
How would it be applying this method also on the expression lists (processAllMatch, processAnyMatch and processNoneMatch) and allow the on_failure on each list (each parent)?

Reply:
All valid expressions (that contain an operator set) are processed by the function on line 83, and this function is used there.